kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.punctuators.action.TaskResult;
+
+@Slf4j
+public abstract class AbstractThrottledPunctuator<T> implements Punctuator {
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
+  private final ThrottledPunctuatorConfig config;
+
+  public AbstractThrottledPunctuator(
+      Clock clock,
+      ThrottledPunctuatorConfig config,
+      KeyValueStore<Long, ArrayList<T>> objectStore) {
+    this.clock = clock;
+    this.config = config;
+    this.objectStore = objectStore;
+  }
+
+  public void scheduleTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    objectsAtWindow.add(object);
+    objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+  }
+
+  public boolean rescheduleTask(long atTimestampInMs, long toTimestampInMs, T object) {
+    if (cancelTask(atTimestampInMs, object)) {
+      scheduleTask(toTimestampInMs, object);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  public boolean cancelTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    boolean removed = objectsAtWindow.remove(object);
+    if (objectsAtWindow.isEmpty()) {
+      objectStore.delete(windowAlignedTimestamp);
+    } else {
+      objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+    }
+    return removed;
+  }
+
+  @Override
+  public final void punctuate(long punctuateTimestamp) {
+    long startTimestamp = clock.millis();
+    log.debug(
+        "Processing tasks with throttling yield of {} until timestamp {}",
+        config.getYieldMs(),
+        punctuateTimestamp);
+    try (KeyValueIterator<Long, ArrayList<T>> it = objectStore.range(0L, punctuateTimestamp)) {
+      while (it.hasNext() && canContinueProcessing(startTimestamp)) {
+        KeyValue<Long, ArrayList<T>> kv = it.next();
+        List<T> objects = kv.value;
+        long windowAlignedTimestamp = kv.key;
+        for (int i = 0; i < objects.size() && canContinueProcessing(startTimestamp); i++) {
+          T object = objects.get(i);
+          TaskResult action = executeTask(punctuateTimestamp, object);
+          if (!cancelTask(windowAlignedTimestamp, object)) {
+            log.debug(
+                "Failed to cancel task at key {}, not found in object store at expected window {}",
+                object,
+                windowAlignedTimestamp);
+          }
+          action
```

Comment:
You need to reschedule before cancel as a good practice. This will guarantee task is in queue in case of error while reschedule or cancel.